db/src/excision.rs

```diff
@@ -134,42 +136,71 @@ pub(crate) fn excisions<'schema>(partition_map: &'schema PartitionMap, schema: &
     }
 }
 
-pub(crate) fn enqueue_pending_excisions(conn: &rusqlite::Connection, schema: &Schema, tx_id: Entid, excisions: ExcisionMap) -> Result<()> {
-    // excisions.into_iter().map(|(entid, excision)| enqueue_pending_excision(self, entid, excision)).collect().and(Ok(()))
+fn excise_datoms(conn: &rusqlite::Connection, excision: &Excision) -> Result<()> {
+    match excision.attrs {
+        Some(ref attrs) => {
+            let s = attrs.iter().join(", ");
+            conn.execute(format!("WITH ids AS (SELECT d.rowid FROM datoms AS d WHERE d.e IS {} AND d.a IN ({})) DELETE FROM datoms WHERE rowid IN ids",
+                                 excision.target, s).as_ref(), &[])?;
+        },
+        None => {
+            conn.execute(format!("WITH ids AS (SELECT d.rowid FROM datoms AS d WHERE (d.e IS {} OR (d.v IS {} AND d.a IS NOT {}))) DELETE FROM datoms WHERE rowid IN ids",
+                                 excision.target, excision.target, entids::DB_EXCISE).as_ref(), &[])?;
+        },
+    }
+
+    Ok(())
+}
+
+fn excise_transactions_before_tx(conn: &rusqlite::Connection, excision: &Excision, before_tx: Entid) -> Result<()> {
+    match excision.attrs {
+        Some(ref attrs) => {
+            let s = attrs.iter().join(", ");
+            conn.execute(format!("WITH ids AS (SELECT t.rowid FROM transactions AS t WHERE t.e IS {} AND t.a IN ({})) DELETE FROM transactions WHERE rowid IN ids",
+                                 excision.target, s).as_ref(), &[])?;
+        },
+        None => {
+            conn.execute(format!("WITH ids AS (SELECT t.rowid FROM transactions AS t WHERE t.tx <= {} AND (t.e IS {} OR (t.v IS {} AND t.a IS NOT {}))) DELETE FROM transactions WHERE rowid IN ids",
+                                 before_tx, excision.target, excision.target, entids::DB_EXCISE).as_ref(), &[])?;
+        },
+    }
 
-    // if !excisions.is_none() {
-    //     bail!(DbError::NotYetImplemented(format!("Excision not yet implemented: {:?}", excisions)));
-    // }
+    Ok(())
+}
 
+pub(crate) fn enqueue_pending_excisions(conn: &rusqlite::Connection, schema: &Schema, tx_id: Entid, excisions: &ExcisionMap) -> Result<()> {
     let mut stmt1: rusqlite::Statement = conn.prepare("INSERT INTO excisions VALUES (?, ?, ?, ?)")?;
     let mut stmt2: rusqlite::Statement = conn.prepare("INSERT INTO excision_attrs VALUES (?, ?)")?;
 
     for (entid, excision) in excisions {
-        stmt1.execute(&[&entid, &excision.target, &excision.before_tx, &excision.before_tx.unwrap_or(tx_id)])?; // XXX
-        if let Some(attrs) = excision.attrs {
-            // println!("attrs {:?}", attrs);
+        let status = excision.before_tx.unwrap_or(tx_id);
```

Comment:
The column name `status` seems a bit misleading and its use lossy. Its use as a "do i need to process anything" means that we don't record the "beforeT" attribute (like we do for other excision parameters like attributes). It _is_ in the transaction table, yes, but the difference in handling seems inconsistent to me.

The "misleading" bit - `status` implies "do i have something to do?", but here its also used to mean "what do i need to do?" at the same time, and the meaning of the column changes as we step through the steps.

I'm guessing you're optimizing away an extra column, essentially?

Reply:
> I'm guessing you're optimizing away an extra column, essentially?

Yes.

I pushed this up before revisiting this with a helpful comment.  You deduced most of the assumptions and asked questions about the rest (great!); but I'll update this comment and add a class comment as well.